File: crates/api/src/handlers/prompt_groups.rs
//! `prompt_groups` domain handlers.
//!
//! Nine IPC commands: five-command CRUD on `prompt_groups` plus four
//! member-management commands (`list_members`, `add_member`,
//! `remove_member`, `set_members`).

use catique_application::{prompt_groups::PromptGroupsUseCase, AppError};
use catique_domain::PromptGroup;
use serde_json::json;
use tauri::State;

use crate::events;
use crate::state::AppState;

/// IPC: list every prompt group.
///
/// # Errors
///
/// Forwards every error from `PromptGroupsUseCase::list`.
#[tauri::command]
pub async fn list_prompt_groups(state: State<'_, AppState>) -> Result<Vec<PromptGroup>, AppError> {
    PromptGroupsUseCase::new(&state.pool).list()
}

/// IPC: look up a prompt group by id.
///
/// # Errors
///
/// Forwards every error from `PromptGroupsUseCase::get`.
#[tauri::command]
pub async fn get_prompt_group(
    state: State<'_, AppState>,
    id: String,
) -> Result<PromptGroup, AppError> {
    PromptGroupsUseCase::new(&state.pool).get(&id)
}

/// IPC: create a prompt group.
///
/// # Errors
///
/// Forwards every error from `PromptGroupsUseCase::create`.
#[tauri::command]
pub async fn create_prompt_group(
    state: State<'_, AppState>,
    name: String,
    color: Option<String>,
    icon: Option<String>,
    position: Option<i64>,
) -> Result<PromptGroup, AppError> {
    let group = PromptGroupsUseCase::new(&state.pool).create(name, color, icon, position)?;
    events::emit(
        &state,
        events::PROMPT_GROUP_CREATED,
        json!({ "id": group.id }),
    );
    Ok(group)
}

/// IPC: partial-update a prompt group.
///
/// # Errors
///
/// Forwards every error from `PromptGroupsUseCase::update`.
#[tauri::command]
pub async fn update_prompt_group(
    state: State<'_, AppState>,
    id: String,
    name: Option<String>,
    color: Option<Option<String>>,
    icon: Option<Option<String>>,
    position: Option<i64>,
) -> Result<PromptGroup, AppError> {
    let group = PromptGroupsUseCase::new(&state.pool).update(id, name, color, icon, position)?;
    events::emit(
        &state,
        events::PROMPT_GROUP_UPDATED,
        json!({ "id": group.id }),
    );
    Ok(group)
}

/// IPC: delete a prompt group.
///
/// # Errors
///
/// Forwards every error from `PromptGroupsUseCase::delete`.
#[tauri::command]
pub async fn delete_prompt_group(state: State<'_, AppState>, id: String) -> Result<(), AppError> {
    PromptGroupsUseCase::new(&state.pool).delete(&id)?;
    events::emit(&state, events::PROMPT_GROUP_DELETED, json!({ "id": id }));
    Ok(())
}

// -------------------------------------------------------------------------
// Member management
// -------------------------------------------------------------------------

/// IPC: list ordered prompt ids for a group.
///
/// # Errors
///
/// Forwards every error from `PromptGroupsUseCase::list_members`.
#[tauri::command]
pub async fn list_prompt_group_members(
    state: State<'_, AppState>,
    group_id: String,
) -> Result<Vec<String>, AppError> {
    PromptGroupsUseCase::new(&state.pool).list_members(&group_id)
}

/// IPC: add a prompt to a group.
///
/// # Errors
///
/// Forwards every error from `PromptGroupsUseCase::add_member`.
#[tauri::command]
pub async fn add_prompt_group_member(
    state: State<'_, AppState>,
    group_id: String,
    prompt_id: String,
    position: i64,
) -> Result<(), AppError> {
    PromptGroupsUseCase::new(&state.pool).add_member(group_id.clone(), prompt_id, position)?;
    events::emit(
        &state,
        events::PROMPT_GROUP_MEMBERS_CHANGED,
        json!({ "group_id": group_id }),
    );
    Ok(())
}

/// IPC: remove a prompt from a group.
///
/// # Errors
///
/// Forwards every error from `PromptGroupsUseCase::remove_member`.
#[tauri::command]
pub async fn remove_prompt_group_member(
    state: State<'_, AppState>,
    group_id: String,
    prompt_id: String,
) -> Result<(), AppError> {
    PromptGroupsUseCase::new(&state.pool).remove_member(group_id.clone(), prompt_id)?;
    events::emit(
        &state,
        events::PROMPT_GROUP_MEMBERS_CHANGED,
        json!({ "group_id": group_id }),
    );
    Ok(())
}

/// IPC: atomically replace the full ordered member list for a group.
///
/// # Errors
///
/// Forwards every error from `PromptGroupsUseCase::set_members`.
#[tauri::command]
pub async fn set_prompt_group_members(
    state: State<'_, AppState>,
    group_id: String,
    ordered_prompt_ids: Vec<String>,
) -> Result<(), AppError> {
    PromptGroupsUseCase::new(&state.pool).set_members(group_id.clone(), ordered_prompt_ids)?;
    events::emit(
        &state,
        events::PROMPT_GROUP_MEMBERS_CHANGED,
        json!({ "group_id": group_id }),
    );
    Ok(())
}

// -------------------------------------------------------------------------
// Group attachment — attach a prompt group as a live unit to a scope.
// Each setter re-materialises the group's current members onto every task
// in scope and emits the scope's `*_UPDATED` event so the frontend
// re-resolves affected task bundles.
// -------------------------------------------------------------------------

/// IPC: set the prompt groups attached to a role.
///
/// # Errors
///
/// Forwards every error from `PromptGroupsUseCase::set_role_groups`.
#[tauri::command]
pub async fn set_role_prompt_groups(
    state: State<'_, AppState>,
    role_id: String,
    group_ids: Vec<String>,
) -> Result<(), AppError> {
    PromptGroupsUseCase::new(&state.pool).set_role_groups(role_id.clone(), group_ids)?;
    events::emit(&state, events::ROLE_UPDATED, json!({ "id": role_id }));
    Ok(())
}

/// IPC: list the prompt groups attached to a role.
///
/// # Errors
///
/// Forwards every error from `PromptGroupsUseCase::list_role_groups`.
#[tauri::command]
pub async fn list_role_prompt_groups(
    state: State<'_, AppState>,
    role_id: String,
) -> Result<Vec<String>, AppError> {
    PromptGroupsUseCase::new(&state.pool).list_role_groups(&role_id)
}

/// IPC: set the prompt groups attached to a board.
///
/// # Errors
///
/// Forwards every error from `PromptGroupsUseCase::set_board_groups`.
#[tauri::command]
pub async fn set_board_prompt_groups(
    state: State<'_, AppState>,
    board_id: String,
    group_ids: Vec<String>,
) -> Result<(), AppError> {
    PromptGroupsUseCase::new(&state.pool).set_board_groups(board_id.clone(), group_ids)?;
    events::emit(&state, events::BOARD_UPDATED, json!({ "id": board_id }));
    Ok(())
}

/// IPC: list the prompt groups attached to a board.
///
/// # Errors
///
/// Forwards every error from `PromptGroupsUseCase::list_board_groups`.
#[tauri::command]
pub async fn list_board_prompt_groups(
    state: State<'_, AppState>,
    board_id: String,
) -> Result<Vec<String>, AppError> {
    PromptGroupsUseCase::new(&state.pool).list_board_groups(&board_id)
}

/// IPC: set the prompt groups attached directly to a task.
///
/// # Errors
///
/// Forwards every error from `PromptGroupsUseCase::set_task_groups`.
#[tauri::command]
pub async fn set_task_prompt_groups(
    state: State<'_, AppState>,
    task_id: String,
    group_ids: Vec<String>,
) -> Result<(), AppError> {
    PromptGroupsUseCase::new(&state.pool).set_task_groups(task_id.clone(), group_ids)?;
    events::emit(&state, events::TASK_UPDATED, json!({ "id": task_id }));
    Ok(())
}

/// IPC: list the prompt groups attached directly to a task.
///
/// # Errors
///
/// Forwards every error from `PromptGroupsUseCase::list_task_groups`.
#[tauri::command]
pub async fn list_task_prompt_groups(
    state: State<'_, AppState>,
    task_id: String,
) -> Result<Vec<String>, AppError> {
    PromptGroupsUseCase::new(&state.pool).list_task_groups(&task_id)
}

File: src/e2e/bridge/handlers/promptGroups.ts
/**
 * Prompt groups command dispatcher (CRUD + membership join).
 */

import type { PromptGroup } from "@bindings/PromptGroup";

import { emitEvent } from "../events";
import { nextId } from "../ids";
import { store } from "../store";
import { nowBig } from "../time";

interface CreatePromptGroupArgs {
  name: string;
  color?: string | null;
  icon?: string | null;
  position?: number;
}

interface UpdatePromptGroupArgs {
  id: string;
  name?: string;
  color?: string | null;
  icon?: string | null;
  position?: number;
}

export function handlePromptGroups(
  command: string,
  args: Record<string, unknown>,
): unknown {
  switch (command) {
    case "list_prompt_groups":
      return Array.from(store.promptGroups.values()).sort((a, b) =>
        Number(a.position - b.position),
      );
    case "get_prompt_group": {
      const id = String(args["id"]);
      const g = store.promptGroups.get(id);
      if (!g) {
        throw {
          kind: "notFound",
          data: { entity: "prompt_group", id },
        };
      }
      return g;
    }
    case "create_prompt_group": {
      const a = args as unknown as CreatePromptGroupArgs;
      const id = nextId("group");
      const ts = nowBig();
      const group: PromptGroup = {
        id,
        name: a.name,
        color: a.color ?? null,
        icon: a.icon ?? null,
        position: BigInt(a.position ?? store.promptGroups.size),
        createdAt: ts,
        updatedAt: ts,
      };
      store.promptGroups.set(id, group);
      store.promptGroupMembers.set(id, []);
      emitEvent("prompt_group:created", { id });
      return group;
    }
    case "update_prompt_group": {
      const a = args as unknown as UpdatePromptGroupArgs;
      const prev = store.promptGroups.get(a.id);
      if (!prev) {
        throw {
          kind: "notFound",
          data: { entity: "prompt_group", id: a.id },
        };
      }
      const next: PromptGroup = {
        ...prev,
        ...(a.name !== undefined ? { name: a.name } : {}),
        ...(a.color !== undefined ? { color: a.color } : {}),
        ...(a.icon !== undefined ? { icon: a.icon } : {}),
        ...(a.position !== undefined ? { position: BigInt(a.position) } : {}),
        updatedAt: nowBig(),
      };
      store.promptGroups.set(a.id, next);
      emitEvent("prompt_group:updated", { id: a.id });
      return next;
    }
    case "delete_prompt_group": {
      const id = String(args["id"]);
      store.promptGroups.delete(id);
      store.promptGroupMembers.delete(id);
      emitEvent("prompt_group:deleted", { id });
      return null;
    }
    case "list_prompt_group_members": {
      const groupId = String(args["groupId"]);
      return store.promptGroupMembers.get(groupId) ?? [];
    }
    case "add_prompt_group_member": {
      const groupId = String(args["groupId"]);
      const promptId = String(args["promptId"]);
      const members = store.promptGroupMembers.get(groupId) ?? [];
      if (!members.includes(promptId)) members.push(promptId);
      store.promptGroupMembers.set(groupId, members);
      emitEvent("prompt_group:members_changed", { group_id: groupId });
      return null;
    }
    case "remove_prompt_group_member": {
      const groupId = String(args["groupId"]);
      const promptId = String(args["promptId"]);
      const members = (store.promptGroupMembers.get(groupId) ?? []).filter(
        (m) => m !== promptId,
      );
      store.promptGroupMembers.set(groupId, members);
      emitEvent("prompt_group:members_changed", { group_id: groupId });
      return null;
    }
    case "set_prompt_group_members": {
      const groupId = String(args["groupId"]);
      const ids = args["orderedPromptIds"] as string[];
      store.promptGroupMembers.set(groupId, [...ids]);
      emitEvent("prompt_group:members_changed", { group_id: groupId });
      return null;
    }
    // ── group attachment (groups as live units) ──────────────────────
    case "list_role_prompt_groups":
      return store.rolePromptGroups.get(String(args["roleId"])) ?? [];
    case "set_role_prompt_groups": {
      const roleId = String(args["roleId"]);
      store.rolePromptGroups.set(roleId, [...(args["groupIds"] as string[])]);
      emitEvent("role:updated", { id: roleId });
      return null;
    }
    case "list_board_prompt_groups":
      return store.boardPromptGroups.get(String(args["boardId"])) ?? [];
    case "set_board_prompt_groups": {
      const boardId = String(args["boardId"]);
      store.boardPromptGroups.set(boardId, [...(args["groupIds"] as string[])]);
      emitEvent("board:updated", { id: boardId });
      return null;
    }
    case "list_task_prompt_groups":
      return store.taskPromptGroups.get(String(args["taskId"])) ?? [];
    case "set_task_prompt_groups": {
      const taskId = String(args["taskId"]);
      store.taskPromptGroups.set(taskId, [...(args["groupIds"] as string[])]);
      emitEvent("task:updated", { id: taskId });
      return null;
    }
    default:
      return undefined;
  }
}
